app/src/main/java/org/mozilla/fenix/components/metrics/AppLaunchTimeMeasurement.kt

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.os.Process
+import android.os.SystemClock
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.HOT
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.WARM
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.COLD
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.UNKNOWN
+import org.mozilla.fenix.perf.Stat
+
+/**
+ * Handles the logic of figuring out launch time for cold, warm and hot startup.
+ */
+class AppLaunchTimeMeasurement(private val stats: Stat = Stat()) {
+
+    private var isOnPreDrawCalled = false
+    private var startupMode = UNKNOWN
+
+    private var applicationOnCreateTimeStampNanoSeconds: Long? = null
+    private var homeActivityInitTimeStampNanoSeconds: Long? = null
+    private var homeActivityOnRestartTimeStampNanoSeconds: Long? = null
+    // we are considering screen to be visible after the first pre draw call.
+    private var homeActivityOnPreDrawTimeStampNanoSeconds: Long = INVALID_TIMESTAMP
+
+    fun onApplicationOnCreate() {
+        startupMode = COLD
+    }
+
+    fun onHomeActivityOnCreate(activityInitNanos: Long) {
+        if (startupMode == UNKNOWN) {
+            startupMode = WARM
+        }
+        this.homeActivityInitTimeStampNanoSeconds = activityInitNanos
+    }
+
+    fun onHomeActivityOnRestart(activityOnRestartNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        startupMode = HOT
+        homeActivityOnRestartTimeStampNanoSeconds = activityOnRestartNanos
+    }
+
+    fun onActivityOnPreDraw(activityOnPreDrawNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        isOnPreDrawCalled = true
+        homeActivityOnPreDrawTimeStampNanoSeconds = activityOnPreDrawNanos
+    }
+
+    /**
+     * if we have both start and finish time for launch, return the difference otherwise return null.
+     */
+    suspend fun getApplicationLaunchTime(): Long? = withContext(Dispatchers.IO) {
+        when {
+            !isOnPreDrawCalled -> {
```

Comment:
When will `onPreDraw` not be called? Should it be better represented in our startup states somehow? Add a comment.

Reply:
since we are sending data onStop(), an example would be launching the app and pressing the home button before onDraw is called.

Reply:
Is that possible? Won't we always draw a frame before `onStop()` is called? e.g.
```
onStart()
onResume()
onDraw() // first frame at least
onPause()
onStop()
```

Or, is it possible that if onStart/onResume take too long, the user can click home and `onDraw` is not called?

Reply:
Spoke on Matrix – `onDraw` isn't called. Can we add a comment to explain this behavior?

>  Should it be better represented in our startup states somehow? 

Not sure how important this is.